packages/next/kotti-field/hooks.ts

```diff
@@ -0,0 +1,282 @@
+	const context = inject<KottiForm.Context | null>(KT_FORM_CONTEXT, null)
+	const translations = useTranslationNamespace('KtFields')
+
+	// sanity-checks
+
+	watch(
+		() => props.formKey,
+		(newFormKey) => {
+			if (context !== null && newFormKey === null)
+				throw new KtFieldErrors.ImplicitFormKeyNone(props)
+
+			if (context === null && newFormKey !== null)
+				throw new KtFieldErrors.InvalidPropOutsideOfContext(props, 'formKey')
+		},
+	)
+
+	watch(
+		() => props.validatorKey,
+		(newValidatorKey) => {
+			if (context === null && newValidatorKey !== null)
+				throw new KtFieldErrors.InvalidPropOutsideOfContext(
+					props,
+					'validatorKey',
+				)
+		},
+	)
+
+	/**
+	 * Some fields do not support a subset of the KtField.Props.
+	 * We explicitly throw errors for props that are not effectively
+	 * consumed (i.e. not to be displayed). This prevents users
+	 * from accidentally passing unneeded or misleading props.
+	 */
+	watchEffect(() => {
+		type Key = keyof KottiField.Supports
+		type Value = Array<keyof KottiField.Props<DATA_TYPE>>
+
+		const PROPS_TO_CHECK_FOR_SUPPORTS: Record<Key, Value> = {
+			clear: ['hideClear'],
+			decoration: ['leftIcon', 'rightIcon', 'prefix', 'suffix'],
+			tabIndex: ['tabIndex'],
+		}
+
+		for (const [supportsKey, propsToCheck] of Object.entries(
+			PROPS_TO_CHECK_FOR_SUPPORTS,
+		) as Array<[Key, Value]>)
+			if (!supports[supportsKey])
+				for (const propKey of propsToCheck)
+					if (props[propKey] !== KOTTI_FIELD_PROPS[propKey].default)
+						throw new KtFieldErrors.UnsupportedProp(props, {
+							supportsKey,
+							propKey,
+							value: props[propKey],
+						})
+	})
+
+	// fetch value
+
+	const currentValue = computed(
+		(): DATA_TYPE => {
+			const value = (() => {
+				if (context === null) return cloneDeep(props.value)
+
+				switch (props.formKey) {
+					case FORM_KEY_NONE:
+						return cloneDeep(props.value)
+
+					case null:
+						throw new KtFieldErrors.ImplicitFormKeyNone(props)
+
+					default:
+						return context.values.value[props.formKey] as DATA_TYPE
+				}
+			})()
+
+			if (!isCorrectDataType(value))
+				throw new KtFieldErrors.InvalidDataType(props, value)
+
+			return value
+		},
+	)
+
+	// formPath
+
+	const formPath = computed(() => {
+		if (context === null) return []
+
+		if (props.formKey === null)
+			throw new KtFieldErrors.ImplicitFormKeyNone(props)
+
+		return props.formKey === FORM_KEY_NONE
+			? [...context.formPath.value]
+			: [...context.formPath.value, props.formKey]
+	})
+
+	// validation
+
+	const isLoading = computed(() => {
+		if (props.isLoading) return true
+		if (context !== null) return context.isLoading.value
+		return false
+	})
+
+	const hideValidation = computed(() =>
+		isLoading.value
+			? true
+			: context === null
+			? false
+			: context.hideValidation.value,
+	)
+
+	const isMissingRequiredField = computed(
+		(): boolean => !props.isOptional && isEmpty(currentValue.value),
+	)
+
+	const validation = computed(
```

Comment:
both the property on the reactive field object this computed value are named `validation`, could you add a differentiation?

Reply:
@mvangent 
> The main suggestion I have left to reconsider is to define functions outside of return object of the component script. This would allow for stronger grouping of logic that belongs together. Defining the functions in the return objects feel a bit Vue 2-y, and from a personal standpoint I think the api should only contain properties that are already defined (more readable).

we are currently discussing this - we r going to look into that refacor :+1: